src/arc_gahp/io_loop_pthread.cpp

```diff
@@ -574,70 +574,47 @@ IOProcess::stdinPipeHandler()
 Worker*
 IOProcess::createNewWorker(void)
 {
-	Worker *new_worker = NULL;
-	new_worker = new Worker(newWorkerId());
+	int new_id = newWorkerId();
+	auto [it, success] = m_workers_list.insert({new_id, Worker(new_id)});
+	ASSERT(success);
+	Worker& new_worker = it->second;
 
 	dprintf (D_FULLDEBUG, "About to start a new thread\n");
 
 	// Create pthread
 	pthread_t thread;
 	if( pthread_create(&thread, NULL,
-				worker_function, (void *)new_worker) !=  0 ) {
+				worker_function, (void *)&new_worker) !=  0 ) {
 		dprintf(D_ALWAYS, "Failed to create a new thread\n");
 
-		delete new_worker;
-		return NULL;
+		m_workers_list.erase(it);
+		return nullptr;
 	}
 
 	// Deatch this thread
 	pthread_detach(thread);
 
-	// Insert a new worker to the map
-	m_workers_list[new_worker->m_id] = new_worker;
 	m_avail_workers_num++;
 
-	dprintf(D_FULLDEBUG, "New Worker[id=%d] is created!\n", new_worker->m_id);
-	return new_worker;
+	dprintf(D_FULLDEBUG, "New Worker[id=%d] is created!\n", new_worker.m_id);
+	return &new_worker;
 }
 
 Worker*
 IOProcess::findFreeWorker(void)
 {
-	for( auto worker : m_workers_list ) {
-
-		if( !worker.second->m_is_doing ) {
-
-			return worker.second;
+	for (auto& [key, worker]: m_workers_list) {
+		if (!worker.m_is_doing) {
+			return &worker;
```

Comment:
How about a `std::find_if` here?

Reply:
Not sure I like how wordy that becomes. Is there a more succinct form that this:
`	auto is_idle = [](std::pair<int, Worker> el) {return !el.second.m_is_doing;};
	auto it = std::find_if(m_workers_list.begin(), m_workers_list.end(), is_idle);
	if (it != m_workers_list.end()) {
		return &it->second;
	} else {
		return nullptr;
	}
`